mycpp/gc_str.cc

```diff
@@ -525,6 +525,15 @@ List<Str*>* Str::split(Str* sep) {
   return this->split(sep, len(this));
 }
 
+int Str::hash(HashFunc h) {
+  // Use the MSB as a flag to indicate that we've already computed the hash.
+  // This keeps the header compact, since we don't have to add a bool.
+  if ((hash_value_ & 0x08000000) == 0) {
```

Comment:
`0x0800_0000` isn't the MSB for 32 bits, it's `0x8000_0000`

That's what I meant by the previous comment :)

while I guess this works, let's use a bitfield to prevent confusion/mistakes

I haven't used bitfields before this project, but I think it will work fine here -- they seem to be very convenient

So we should have something like

```
unsigned hash : 31
unsigned is_hashed : 1
```

Reply:
OH lol! Not sure how those got swapped... Nice eye. I guess this works because any bit is fine as long as we're consistently use the same one. Will fix.

Bitfields are somewhat annoying, because you have to guard them with a check on `__BYTE_ORDER` and define the layout twice, like below. Using the value instead of the layout of the int is a way to avoid this, but I get why it might seem more cumbersome. Are you willing to relax the "minimize ifdefs" goal for this?
```
#if __BYTE_ORDER == __BIG_ENDIAN
unsigned hash : 31
unsigned is_hashed : 1
#elif __BYTE_ORDER == __LITTLE_ENDIAN
unsigned is_hashed : 1
unsigned hash : 31
#else
// error
#endif
```

Reply:
I think you only need the `#ifdef` if you're sending it over the network, which we aren't doing

The other case you might need it is if you overlap integers/pointers, assuming they will line up, which we're not doing here

I think this is a "pure" usage of bitfields -- we have a 31 bit integer and a 1-bit int, and no other constraints

Only constraint is size

Reply:
Cool. Will tweak this